pandas/indexes/category.py

```diff
@@ -574,6 +574,32 @@ def append(self, other):
         codes = np.concatenate([c.codes for c in to_concat])
         return self._create_from_codes(codes, name=name)
 
+    def _array_putmask(self, a, mask, values):
+        """
+        Equivalent to np.putmask(a, mask, values).
+
+        "a.categories" must already contain all categories that "values" uses.
+        """
+
+        rep_vals = a.categories.get_indexer(values.categories
+                                            ).view(dtype=a.codes.dtype)
+
+        idces = np.digitize(values.codes, np.arange(len(values.categories)),
```

Comment:
see ``union_categoricals``.  I haven't really looked at this, but I don't see any reason to have new categories at all. That violates the guarantee. you would use ``pd.factorize`` instead anyhow. But you should put code in approrpriate places.

I don't have much time to look into this.